presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -0,0 +1,197 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner;
+
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.metadata.Signature;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeSignature;
+import com.facebook.presto.sql.analyzer.Analysis;
+import com.facebook.presto.sql.tree.ArrayConstructor;
+import com.facebook.presto.sql.tree.Cast;
+import com.facebook.presto.sql.tree.Cube;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.GroupingElement;
+import com.facebook.presto.sql.tree.GroupingOperation;
+import com.facebook.presto.sql.tree.GroupingSets;
+import com.facebook.presto.sql.tree.LongLiteral;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.QuerySpecification;
+import com.facebook.presto.sql.tree.Rollup;
+import com.facebook.presto.sql.tree.SortItem;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.facebook.presto.type.ListLiteralType;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Arrays;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.operator.scalar.GroupingOperationFunction.GROUPING;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.spi.type.IntegerType.INTEGER;
+import static com.facebook.presto.sql.analyzer.ExpressionAnalyzer.resolveFunction;
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class GroupingOperationRewriter
+        extends ExpressionRewriter<Void>
+{
+    private final QuerySpecification queryNode;
+    private final Analysis analysis;
+    private final Metadata metadata;
+    private boolean containsGroupingOperation = false;
+
+    public GroupingOperationRewriter(QuerySpecification queryNode, Analysis analysis, Metadata metadata)
+    {
+        this.queryNode = requireNonNull(queryNode, "node is null");
+        this.analysis = requireNonNull(analysis, "analysis is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+    }
+
+    @Override
+    public Expression rewriteExpression(Expression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return treeRewriter.defaultRewrite(node, context);
+    }
+
+    @Override
+    public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        FunctionCall functionCall = treeRewriter.defaultRewrite(node, context);
+
+        if (functionCall != node && containsGroupingOperation) {
```

Comment:
When some other rewrite is performed. Since this method is overriding the one in `ExpressionRewriter`, other `FunctionCall` objects that are rewritten with this will go through this method and we only want to update types for grouping functions. Let me know if this is unclear and I can explain in person.